components/fxa-client/src/pairing_channel/client_states.rs

```diff
@@ -0,0 +1,388 @@
+// that provide different behaviours of `send` and `receive` depending
+// on the state of the connection.
+//
+pub enum ClientState<F: FnMut(&[u8]) -> Result<()>> {
+    Invalid,
+    ClientWaitEe(ClientWaitEe<F>),
+    ClientWaitSh(ClientWaitSh<F>),
+    ClientWaitFinished(ClientWaitFinished<F>),
+    ClientConnected(ClientConnected<F>),
+}
+
+impl<F: FnMut(&[u8]) -> Result<()>> ClientState<F> {
+    #[allow(dead_code)]
+    pub fn client_init(psk: Vec<u8>, psk_id: Vec<u8>, callback: F) -> Result<Self> {
+        let client_start = ClientStart {
+            key_schedule: KeySchedule::new(),
+            record_layer: RecordLayer::new(callback),
+        };
+        let client_start = client_start.initialize(&psk, &psk_id)?;
+        Ok(ClientState::ClientWaitSh(client_start))
+    }
+
+    pub fn recv(&mut self, data: &[u8]) -> Result<(RecordType, Vec<u8>)> {
+        match self {
+            ClientState::ClientWaitEe(inner) => inner.record_layer.recv(data),
+            ClientState::ClientWaitSh(inner) => inner.record_layer.recv(data),
+            ClientState::ClientWaitFinished(inner) => inner.record_layer.recv(data),
+            ClientState::ClientConnected(inner) => inner.record_layer.recv(data),
+            _ => anyhow::bail!("Invalid state"),
+        }
+    }
+
+    pub fn recv_application_data(&mut self, data: &[u8]) -> Result<Vec<u8>> {
+        match self {
+            ClientState::ClientConnected(inner) => Ok(inner.recv_application_data(data)),
+            _ => anyhow::bail!("State not connected"),
+        }
+    }
+
+    pub fn recv_change_cipher_spec(&mut self, data: &[u8]) -> Result<()> {
+        match self {
+            ClientState::ClientWaitEe(inner) => inner.recv_change_cipher_spec(data),
+            _ => anyhow::bail!("Invalid state"),
+        }
+    }
+
+    pub fn add_to_transcript(&mut self, data: &[u8]) -> Result<()> {
+        match self {
+            ClientState::ClientWaitEe(inner) => inner.key_schedule.add_to_transcript(data),
+            ClientState::ClientWaitSh(inner) => inner.key_schedule.add_to_transcript(data),
+            ClientState::ClientWaitFinished(inner) => inner.key_schedule.add_to_transcript(data),
+            ClientState::ClientConnected(inner) => inner.key_schedule.add_to_transcript(data),
+            _ => anyhow::bail!("Invalid state"),
+        };
+        Ok(())
+    }
+
+    pub fn recv_handshake_message(self, message_bytes: &[u8]) -> Result<Self> {
+        let msg = HandshakeMessage::from_bytes(message_bytes)?;
+        match self {
+            ClientState::ClientWaitSh(inner) => Ok(ClientState::ClientWaitEe(
+                inner.recv_handshake_message(msg)?,
+            )),
+            ClientState::ClientWaitEe(inner) => Ok(ClientState::ClientWaitFinished(
+                inner.recv_handshake_message(msg)?,
+            )),
+            ClientState::ClientWaitFinished(inner) => Ok(ClientState::ClientConnected(
+                inner.recv_handshake_message(msg)?,
```

Comment:
I'm lightly surprised that you have to repeat the fact that this produces a `ClientConnected` twice - once down in the type signature of `ClientWaitFinished::recv_handeshake_message`, and again here when you wrap it into the top-level enum. It's not a problem per se, more of a vague "I wonder if you're fighting against the abstraction layer" kind of thing. 

Does the outer layer here really need to be an `enum`, allowing calling code to inspect what the current state is?

An alternative factoring could be:

* Define a `ClientState` trait with all the methods like `recv_handshake_message` on it.
* Have each of `ClientStart`, `ClientWaitServerHello` etc independently impl the `ClientState` trait.
* Have a top-level object to encapsulate the current state, but it's opaque instead of an enum:
    * `inner` could be a `Box<dyn ClientState>` or similar to hold the current state.
    * calling `recv_handshake_message` on the top-level object would call it on `inner`, than potentially change `inner` based on the result

This would be roughtly the rust equivalent of having an abstract `ClientState` base class, and defining a concrete subclass for each of the different states.

Might make things better, might make things worse, but I think it would at least reduce a little bit of enum-matching boilerplate here. YMMV.

Reply:
Hmmm Ya thought about that very briefly (Since that's how it's done in the js implementation), the reason I ended up this direction was an attempt to keep the transitions explicit and so that I don't need to implement functions for States that shouldn't have them. Although now that I think about it, it's possible to keep the explicit types and have the `Box<dyn ClientState>`. I'll experiment with it! Thanks Ryan!